src/condor_starter.V6.1/starter_v61_main.cpp

```diff
@@ -229,6 +229,81 @@ printClassAd( void )
 	if (!method_list.empty()) {
 		printf("%s = \"%s\"\n", ATTR_HAS_FILE_TRANSFER_PLUGIN_METHODS, method_list.c_str());
 	}
+
+	// even if we have no transfer plugin methods, we may want to
+	// advertise plugin extra attributes
+	if ( ! ft.getPluginQueryAds().empty()) {
+		// publish plugin extra attributes, GetSupportedMethods will collect plugin ads
+		// when it queries the plugins for their '-classad' 
+		// We will incorporate into the starter classad any attributes from the plugins
+		// that are marked as SlotAttrs or StartdAttrs
+		ClassAd extraAd;
+		classad::References failed_methods;
+		classad::References startd_attrs;
+		// when SlotAttrs are enabled, a plugin choose whether to put an attribute
+		// into the daemon ad, or into the slot ads. otherwise all attributes go into all ads
+	   #ifdef ENABLE_PLUGIN_SLOT_ATTRS
+		classad::References slot_attrs;
+	   #endif
+		std::string listval;
+
+		for (auto & ad : ft.getPluginQueryAds()) {
+			classad::References has_attrs;
+			if (ad.LookupString("FailedMethods", listval)) {
+				for (auto & meth : StringTokenIterator(listval)) { failed_methods.insert(meth); }
+			}
+			if (ad.LookupString("StartdAttrs", listval)) {
+				for (auto & attr : StringTokenIterator(listval)) {
+					classad::Value val;
+					if (ad.EvaluateAttr(attr, val)) { // this returns only SAFE values, so no lists or classad
+						extraAd.Insert(attr, classad::Literal::MakeLiteral(val));
+						startd_attrs.emplace(attr);
+						has_attrs.emplace(attr);
+					}
+				}
+			}
+		   #ifdef ENABLE_PLUGIN_SLOT_ATTRS
+			if (ad.LookupString("SlotAttrs", listval)) {
+				for (auto & attr : StringTokenIterator(listval.c_str())) {
+					if ( ! has_attrs.count(attr)) {
+						classad::Value val;
+						if ( ! ad.EvaluateAttr(attr, val)) { // this returns only SAFE values, so no lists or classad
+							continue;
+						} 
+						extraAd.Insert(attr, classad::Literal::MakeLiteral(val));
+					}
+					slot_attrs.emplace(attr);
+				}
+			}
+		   #endif
+		}
+	   #ifdef ENABLE_PLUGIN_SLOT_ATTRS
+		if ( ! slot_attrs.empty()) {
+			listval = JoinAttrNames(slot_attrs, " ");
+			extraAd.Assign("SlotAttrs", listval);
+		}
+		if ( ! startd_attrs.empty()) {
+			listval = JoinAttrNames(startd_attrs, " ");
+			extraAd.Assign("StartdAttrs", listval);
+		}
+	   #endif
```

Comment:
Shouldn't the ``#endif`` be before the if conditional checking if the ``startd_attrs`` is empty?

Reply:
no.  if we aren't distinguishing between slot attrs and startd attrs, we don't need either one.